storage/src/main/scala/coop/rchain/storage/package.scala

```diff
@@ -139,34 +154,34 @@ package object storage {
       implicit m: Match[P, A]): Option[(K, List[A])] =
     store.withTxn(store.createTxnWrite()) { txn =>
       val groupedChannels: List[List[C]] = store.getJoin(txn, channel)
-      store.putA(txn, List(channel), Datum(data, persist))
-      logger.info(s"produce: persisted <data: $data> at <channel: $channel>")
       logger.info(
         s"produce: searching for matching continuations at <groupedChannels: $groupedChannels>")
-      extractProduceCandidate(store, groupedChannels)(txn)
-        .map {
-          case ProduceCandidate(channels,
-                                WaitingContinuation(_, continuation, persistK),
-                                continuationIndex,
-                                dataCandidates) =>
-            if (!persistK) {
-              store.removePsK(txn, channels, continuationIndex)
-            }
-            dataCandidates.foreach {
-              case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex)
-                  if !persistData =>
+      extractProduceCandidateAlt(store, groupedChannels, channel, Datum(data, persist))(txn) match {
+        case Some(
+            ProduceCandidate(channels,
+                             WaitingContinuation(_, continuation, persistK),
+                             continuationIndex,
+                             dataCandidates)) =>
+          if (!persistK) {
+            store.removePsK(txn, channels, continuationIndex)
+          }
+          dataCandidates.foreach {
+            case DataCandidate(candidateChannel, Datum(_, persistData), dataIndex) =>
+              if (!persistData && dataIndex >= 0) {
                 store.removeA(txn, candidateChannel, dataIndex)
-                ignore { store.removeJoin(txn, candidateChannel, channels) }
-              case _ =>
-                ()
-            }
-            logger.info(s"produce: matching continuation found at <channels: $channels>")
-            (continuation, dataCandidates.map(_.datum.a))
-        }
-        .recoverWith {
-          case _: Unit =>
-            logger.info(s"produce: no matching continuation found")
-            None
-        }
+              }
+              store.removeJoin(txn, candidateChannel, channels)
```

Comment:
Is this correct?  

The tests seem to corroborate that it is.

Reply:
Yes. It's correct.